File: practice/old_solutions/pcycle.pl
#
# Problem: Permutation Cycles
# URL: http://www.codechef.com/problems/PCYCLE
#
# Author: Ronald Kaiser
# Email: raios dot catodicos at gmail dot com
#


use feature qw(say);

chop(my $n = <>);
my @list = (-1, split(/\s+/, <>));

my @m = (-1, ((0)x$n));

my @current_cycle = ();
my @all_cycles = ();
my $i = 1;

while (1) {
    push @current_cycle, $i;
    if ($m[$i] == 1) {
        my @copy = @current_cycle;
        @all_cycles = (@all_cycles, \@copy);
        @current_cycle = ();
        my @r = grep $m[$_] == 0, 1..$#m;
        if (@r > 0) {
             $i = shift @r;
        }
        else {
            say scalar(@all_cycles);
            foreach my $cycle (@all_cycles) {
                say "@{$cycle}";
            }
            exit 0;
        };
    }
    else {
        $m[$i] = 1;
        $i = $list[$i];
    }
}

File: practice/easy/PCYCLE.py
"""
  Url: https://www.codechef.com/problems/PCYCLE
"""


__author__ = "Ronald Kaiser"
__email__ = "raios dot catodicos at gmail dot com"


N = int(input())
ns = [-1] + list(map(int, input().split()))

start = i = 1
cycles = []
cur_cycle = []
m = [-1] + [0]*(N)
count = N

while True:
    if i == start and m[i] == 1:
        cur_cycle.append(i)
        cycles.append(cur_cycle)
        cur_cycle = []
        if count is not 0:
            start = i = m.index(0)
            continue
        break
    else:
        cur_cycle.append(i)
        i = ns[i]
        m[i] = 1
        count -= 1

print(len(cycles))
for c in cycles:
    print(' '.join(map(str, c)))
